File: libfgg/pool/pool_test.go
package pool

import (
	"os"
	"path"
	"testing"
	"time"
)

func TestPool(t *testing.T) {
	// 100M
	sendPath, _, err := HelpCreateTmpFile("filegogo_libfgg_pool_send", 100*1024*1024)
	if err != nil {
		t.Fatal(err)
	}

	defer os.Remove(sendPath)

	recvPath := path.Join(os.TempDir(), "filegogo_libfgg_pool_recv")
	defer os.Remove(recvPath)

	sender := New()
	if err := sender.SetSend(sendPath); err != nil {
		t.Error(err)
	}

	recver := New()
	if err := recver.SetRecv(recvPath); err != nil {
		t.Error(err)
	}

	meta, err := sender.SendMeta()
	if err != nil {
		t.Error(err)
	}

	if err := recver.RecvMeta(meta); err != nil {
		t.Error(err)
	}

	running := true
	recver.OnFinish = func() {
		running = false
	}
	for running {
		c := recver.Next()
		if c == nil {
			time.Sleep(100 * time.Millisecond)
		} else {
			data, err := sender.SendData(c)
			if err != nil {
				t.Error(err)
			}

			if err := recver.RecvData(c, data); err != nil {
				t.Error(err)
			}
		}
	}

	hash, err := sender.SendHash()
	if err != nil {
		t.Error(err)
	}

	if err := recver.RecvHash(hash); err != nil {
		t.Error(err)
	}
}

File: webapp/libfgg/pool/pool.test.ts
import { assert, describe, it } from 'vitest'

import VirtualFile from "./file/virtual"

import Pool from "./pool"

describe('file pool test', async () => {
  const nameSend = "send_xxx"
  const size = 1024*1024

  const fileSend = new VirtualFile(nameSend, size)
  const fileRecv = new VirtualFile("recv_xxx", size)

  const sender = new Pool
  sender.setSend(fileSend)

  const meta = await sender.sendMeta()

  it('meta file name', () => {
    assert.equal(nameSend, meta.name)
  })

  it('meta file size', () => {
    assert.equal(size, meta.size)
  })

  const recver = new Pool

  recver.setRecv(fileRecv)
  recver.recvMeta(meta)

  const c = recver.next()

  it('recv next data chunk', () => {
    assert.notEqual(c, null)
  })

  if (c) {
    const data = await sender.sendData(c)
    await recver.recvData(c, data)
  }

  const run = async (): Promise<void> => {
    return new Promise((resolve) => {
      const timer = setInterval(async() => {
        const c = recver.next()
        if (c) {
          const data = await sender.sendData(c)
          await recver.recvData(c, data)
        }
      }, 100)

      recver.OnFinish = () => {
        clearInterval(timer)
        resolve()
      }
    })
  }

  await run()

  const hash = sender.sendHash()
  const ok = recver.recvHash(hash)

  it('check sum', () => {
    assert.equal(ok, true)
  })
})
